### ISSUE TITLE
Make it possible to delete parent resources

### ISSUE THREAD

**Issue by maljub01** [metacontroller/metacontroller#77]
I'm using metacontroller to implement a resource that works similarly to batch jobs. This works fine at first, but after a while the system starts to accumulate a lot of those "finished" jobs. I'd like to be able to implement something like `ttlSecondsAfterFinished`<sup>1</sup> for my resource using metacontroller.

As far as I can tell, this is not currently possible. However, IIUC it should be relatively easy to add a `delete` flag to the data returned by the controller which would be used for "clean up" purposes.

I'm happy to volunteer my time to implement this, or an alternative solution if you prefer.

[1] See: https://kubernetes.io/docs/concepts/workloads/controllers/job/#clean-up-finished-jobs-automatically

**Comment by grzesuav** [comment]
thanks for input. The problem I see with `delete` is that if the resource is children resource for some other controller, it might lead to some faulty behavior. Not sure also about events versioning in k8s (i.e. if we can get two subsequent updates from k8s which are out of order).


Calling @AmitKumarDas // @enisoc for comments

**Comment by grzesuav** [comment]
Another view on that case - maybe we can use `Status` field to mark that metacontroller finished working with parent resource, and then some garbage-collector mechanism on cluster can prune it ?

**Comment by enisoc** [comment]
> However, IIUC it should be relatively easy to add a delete flag to the data returned by the controller which would be used for "clean up" purposes.

That makes sense to me.

> if the resource is children resource for some other controller, it might lead to some faulty behavior

Perhaps we can reject the deletion request if there are any controllerRefs. It sounds like the use case here is to clean up "top-level" objects that are not managed by any controller.

> Not sure also about events versioning in k8s (i.e. if we can get two subsequent updates from k8s which are out of order).

Watch events shouldn't be out of order, but they might be out of date which could cause us to delete something by mistake. I think there is now a ResourceVersion precondition available, although I haven't used it before. I think the feature was still under development back when I last looked:

https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Preconditions

If that feature is fully functional now, it should let us send a Delete request that will only succeed if the object version matches what we sent to the hook. If the deletion fails due to a version mismatch, we can wait for the new version to come in as a watch event, at which point we'll send another hook request to see if the user still wants us to delete the object.

**Comment by fpetkovski** [comment]
I wonder if it is worth considering implementing a generic TTL controller which can delete resources based on a `ttl` and a status field (such as a `Condition`). This will decouple the deletion from the existing controllers and will make such a TTL controller composable and independent.

This mechanism is also in line with how ttl is implemented for job resources: https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/. 
Unfortunately, the built in TTL controller only works for Jobs now.

A generic TTL controller definition from a user's perspective would look something like this:
```
apiVersion: metacontroller.k8s.io/v1alpha1
kind: TTLController
metadata:
  name: ttl-controller-for-my-resource
spec:
  resources:
    - apiVersion: myApiVersion/v1
       resource: myResource
       selector:
         somekey: somevalue 
       ttlSeconds: 120
       completionField: status.CompletionTimestamp
```

**Comment by AmitKumarDas** [comment]
@fpetkovski Do you mind trying `GenericController` from metac & see if it suits your usecase. We might need to tune it to solve your problem. If that works out, you might perhaps want to get it merged into metacontroller as well :)

schema - _https://github.com/AmitKumarDas/metac/blob/master/apis/metacontroller/v1alpha1/types_generic.go_

examples - _https://github.com/AmitKumarDas/metac/tree/master/examples/gctl_

_NOTE: GenericController can work in CRD mode or just as a plain config file mode._

**Comment by grzesuav** [comment]
> I wonder if it is worth considering implementing a generic TTL controller which can delete resources based on a `ttl` and a status field (such as a `Condition`). This will decouple the deletion from the existing controllers and will make such a TTL controller composable and independent.

Good idea, however still not sure it can be done in metacontroller in current design, parent resources are not modifiable.


I guess we would need to add additional feature considering Enisoc thoughts.

**Comment by maljub01** [comment]
My only problem with the generic TTL controller suggestion is that it feels too tailored for solving the "implement TTL" issue, and that metacontroller users interested in managing deletion would prefer to have more control over when to trigger it.

For example, one thing I'd like to experiment with is having different retention schedules depending on the final status of the object/job.

**Comment by fpetkovski** [comment]
> @fpetkovski Do you mind trying `GenericController` from metac & see if it suits your usecase. We might need to tune it to solve your problem. If that works out, you might perhaps want to get it merged into metacontroller as well :)

Will take a look, thanks!



> My only problem with the generic TTL controller suggestion is that it feels too tailored for solving the "implement TTL" issue, and that metacontroller users interested in managing deletion would prefer to have more control over when to trigger it.
> 
> For example, one thing I'd like to experiment with is having different retention schedules depending on the final status of the object/job.

Would that be solved by adding a `statusSelector` which you can use to filter out resources which meet a certain criteria? Since you would control both the `statusSelector` and the resources' status, you would be able to have different retention periods depending on the outcome. For example, you can implement different policies for successful and failed objects this way. Or do you think this is still not flexible enough?

**Comment by ferozed** [comment]
I would like to vote up this issue, and also add another scenario where this is needed.

We are using kubernetes CRD to create external resources ( hivetables ). The resource creation is initiated, not through a `kubectl apply -f <filename>` command, but using terraform. We wrote a terraform provider that can talk to metacontroller to create the resource, the request for the creation is then forwarded to the controller.

The controller does validation of the request before creating the external HiveTable resource. When the validation fails, we find that the CRD still exists. This is not a good thing because it will result in a lot of CRDs in a bad state, and the controller constantly getting `sync` requests for these resources that are always going to fail the same way (validation error).

So for this scenario, I think @maljub01 original suggestion of having the controller return a special `delete` flag as part of the `status` or some other field, would be a good option.
